src/coreclr/jit/promotiondecomposition.cpp

```diff
@@ -24,13 +25,1181 @@ class DecompositionStatementList
+                return m_compiler->gtNewZeroConNode(type);
+            }
+            default:
+                unreached();
+        }
+    }
+
+    //------------------------------------------------------------------------
+    // FinalizeCopy:
+    //   Create IR to perform the decomposed copy.
+    //
+    // Parameters:
+    //   statements - List to add statements to.
+    //
+    void FinalizeCopy(DecompositionStatementList* statements)
+    {
+        assert(m_dst->OperIs(GT_LCL_VAR, GT_LCL_FLD, GT_BLK) && m_src->OperIs(GT_LCL_VAR, GT_LCL_FLD, GT_BLK));
+
+        RemainderStrategy remainderStrategy = DetermineRemainderStrategy();
+
+        // If the remainder is a full block and is going to incur write barrier
+        // then avoid incurring multiple write barriers for each source
+        // replacement that is a GC pointer -- write them back to the struct
+        // first instead.
+        if ((remainderStrategy.Type == RemainderStrategy::FullBlock) && m_dst->OperIs(GT_BLK) &&
+            m_dst->GetLayout(m_compiler)->HasGCPtr())
+        {
+            for (int i = 0; i < m_entries.Height(); i++)
+            {
+                const Entry& entry = m_entries.BottomRef(i);
+                // TODO: Double check that TYP_BYREF do not incur any write barriers.
+                if ((entry.FromReplacement != nullptr) && (entry.Type == TYP_REF))
+                {
+                    Replacement* rep = entry.FromReplacement;
+                    if (rep->NeedsWriteBack)
+                    {
+                        statements->AddStatement(
+                            Promotion::CreateWriteBack(m_compiler, m_src->AsLclVarCommon()->GetLclNum(), *rep));
+                        JITDUMP("  Will write back V%02u (%s) to avoid an additional write barrier\n", rep->LclNum,
+                                rep->Description);
+
+                        rep->NeedsWriteBack = false;
+                    }
+                }
+            }
+        }
+
+        GenTree*     addr       = nullptr;
+        GenTreeFlags indirFlags = GTF_EMPTY;
+
+        if (m_dst->OperIs(GT_BLK))
+        {
+            addr = m_dst->gtGetOp1();
+            indirFlags =
+                m_dst->gtFlags & (GTF_IND_VOLATILE | GTF_IND_NONFAULTING | GTF_IND_UNALIGNED | GTF_IND_INITCLASS);
+        }
+        else if (m_src->OperIs(GT_BLK))
+        {
+            addr = m_src->gtGetOp1();
+            indirFlags =
+                m_src->gtFlags & (GTF_IND_VOLATILE | GTF_IND_NONFAULTING | GTF_IND_UNALIGNED | GTF_IND_INITCLASS);
+        }
+
+        int numAddrUses = 0;
+
+        if (addr != nullptr)
+        {
+            for (int i = 0; i < m_entries.Height(); i++)
+            {
+                if (!IsHandledByRemainder(m_entries.BottomRef(i), remainderStrategy))
+                {
+                    numAddrUses++;
+                }
+            }
+
+            if (remainderStrategy.Type != RemainderStrategy::NoRemainder)
```

Comment:
Hmm, it would probably be reasonable to try to make this more polymorphic, but let me leave it as is for now and see how (if) it ends up abstracting/generalizing in the future to more sophisticated strategies.